cjdns-core/src/encoding.rs

```diff
@@ -31,393 +32,614 @@
+        InvalidFormsAmount,
+
+        /// Scheme with single form must not have non-empty prefix value
+        #[error("Invalid encoding scheme: single form has non-empty prefix")]
+        SingleFormWithPrefix,
+
+        /// Scheme with multiple forms must have non-empty prefix value
+        #[error("Invalid encoding scheme: multiple forms - prefix length is out of bounds (1..32)")]
+        MultiFormBadPrefix,
 
-    #[test]
-    fn test_multiple_forms() {
-        // { bitCount: 4, prefix: "01", prefixLen: 1 },
-        // { bitCount: 8, prefix: "00", prefixLen: 1 },
-        // 81        0c        08
-        // 1000 0001 0000 1100 0000 1000
-        let mut input = [
-            EncodingSchemeForm { bit_count: 4, prefix_len: 1, prefix: 1 },
-            EncodingSchemeForm { bit_count: 8, prefix_len: 1, prefix: 0 },
-        ].to_vec();
-
-        let mut serialized = serialize_forms(&input.to_vec()).expect("failed to serialize");
-        // https://github.com/cjdelisle/cjdnsencode/blob/89216230daa82eb43689c6af48de3c6a138002f1/test.js#L21
-        assert_eq!(serialized, [0x81, 0x0c, 0x08].to_vec());
-        let mut deserialized = deserialize_forms(&serialized).expect("failed to deserialize");
-        assert_eq!(deserialized, input);
-        assert!(validate(&deserialized).is_ok());
-
-        // name: "SCHEME_v358",
-        // obj: [
-        //   { bitCount: 3, prefix: "01", prefixLen: 1 },
-        //   { bitCount: 5, prefix: "02", prefixLen: 2 },
-        //   { bitCount: 8, prefix: "00", prefixLen: 2 }
-        // ],
-        // hex: '6114458100'
-        // 61        14        45        81  :      00
-        // 0110 0001 0001 0100 0100 0101 1000 0001 0000 0000
-        input = [
-            EncodingSchemeForm { bit_count: 3, prefix_len: 1, prefix: 1 },
-            EncodingSchemeForm { bit_count: 5, prefix_len: 2, prefix: 2 },
-            EncodingSchemeForm { bit_count: 8, prefix_len: 2, prefix: 0 },
-        ].to_vec();
-
-        serialized = serialize_forms(&input.to_vec()).expect("failed to serialize");
-        // https://github.com/cjdelisle/cjdnsencode/blob/89216230daa82eb43689c6af48de3c6a138002f1/test.js#L30
-        assert_eq!(serialized, [0x61, 0x14, 0x45, 0x81, 0x0].to_vec());
-        deserialized = deserialize_forms(&serialized).expect("failed to deserialize");
-        assert_eq!(deserialized, input);
-        assert!(validate(&deserialized).is_ok());
+        /// Multiple forms should have `bit_count` in ascending order
+        #[error("Invalid encoding scheme: multiple forms should have bit_count in ascending order")]
+        BitCountNotSorted,
+
+        /// Multiple forms must have unique prefixes
+        #[error("Invalid encoding scheme: multiple forms must have unique prefixes")]
+        DuplicatePrefix,
+
+        /// Form bit size should be less than 59 for multiple form schemes
+        #[error("Invalid encoding scheme: form size too big (bit_count + prefix_len > 59)")]
```

Comment:
This error message is not informative at all, and by the way it is different from the original message in JS code.
No need to change proper message with a silly one.
Use this: "Encoding scheme cannot be represented in the usable space in a 64-bit label".

Reply:
Also it makes sense to turn 59 into a named const, with a comment like this:

> In the old days every label needed to be topped with `0001`. To make sure that no label would ever go over 64 bits even with `0001` spliced on the top of it, we use this reservation.
>
> Now that we have a `shift` field in the header, that's all but obsolete, a label can go up to bit `64` if needed, as long as the final bit is a `1`.

Reply:
507640d